### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by bhawanaprasain**
Users can click "Show First Run Wizard Again" button in General section of Settings to get firstrunwizard popup.

### CODE REVIEW

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 1]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
 		$this->firstRunWizardPage->closePopup();
 	}
 
+	/**
+	 * @When the user clicks :arg1 button
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: $arg is never used

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 2]
```diff
@@ -47,11 +48,22 @@ class WebUIFirstrunwizardContext extends RawMinkContext implements Context {
 	 */
 	private $firstrunwizardPage;
 
+	/**
+	 * @var PathofWizardFileFromServerRoot
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	 * @var string
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 3]
```diff
@@ -47,11 +48,22 @@ class WebUIFirstrunwizardContext extends RawMinkContext implements Context {
...
+	 */
+	private $pathOfWizardFileFromServerRoot = "/apps/firstrunwizard/templates/wizard.php";
+
+	/**
+	 * @var FileContent
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	 * @var string
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 4]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
...
+	public function showFirstRunWizardAgainButton($arg1) {
+		$this->firstRunWizardPage->clickShowFirstRunWizardPopupAgain();
+	}
+
+	/**
+	 * @Then the heading of the popup should be :expectedMessage
+	 *
+	 * @param $expectedMessage
+	 *
+	 * @Throws Exception
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	 * @throws Exception
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 5]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
...
+	 *
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 6]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
...
+		$headingMessage = $this->firstRunWizardPage->getHeadingMessage();
+
+		PHPUnit\Framework\Assert::assertSame(
+			$expectedMessage, $headingMessage, "Firstrunwizard showed unexpected message."
+		);
+	}
+
+	/**
+	 *@Given the administrator has changed the default popup message of firstrunwizard to :newMessage
+	 *
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: @param?

File: tests/acceptance/features/webUIFirstrunwizard/headingMessage.feature [hunk 7]
```diff
@@ -0,0 +1,24 @@
+@webUI @insulated @disablePreviews
+Feature: default message from firstrunwizard is displayed after changing default popup message of firstrunwizard and reverting it back
+  As an administrator
+  I want to check if firstrunwizard displays default message on popup after changing the default message and reverting it
+  So that I can manipulate and send popup message from firstrunwizard to new users
+
+  Scenario: User logs into ownCloud for the first time
```
Comment (dpakach) [inline tests/acceptance/features/webUIFirstrunwizard/headingMessage.feature]: This scenario is same as `tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature:7`

File: tests/acceptance/features/webUIFirstrunwizard/headingMessage.feature [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    And user "user1" has been created with default attributes
+    When user "user1" logs in using the webUI
+    Then the user should see the firstrunwizard popup message
+    And the heading of the popup should be "next default message"
+
+  Scenario: Administrator reverts the popup message of firstrunwizard to default message
+    Given user "user1" has been created with default attributes
+    When user "user1" logs in using the webUI
+    Then the user should see the firstrunwizard popup message
+    And the heading of the popup should be "A safe home for all your data"
```
Comment (dpakach) [inline tests/acceptance/features/webUIFirstrunwizard/headingMessage.feature]: this is also same scenario as above

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 9]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
...
+	 * @Throws Exception
+	 *
+	 * @return void
+	 */
+	public function theHeadingOfThePopupShouldBe($expectedMessage) {
+		$headingMessage = $this->firstRunWizardPage->getHeadingMessage();
+
+		PHPUnit\Framework\Assert::assertSame(
+			$expectedMessage, $headingMessage, "Firstrunwizard showed unexpected message."
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
			$expectedMessage, $headingMessage, "Firstrunwizard was expected to show $expectedMessage but showed $headingMessage."
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 10]
```diff
@@ -47,11 +48,22 @@ class WebUIFirstrunwizardContext extends RawMinkContext implements Context {
...
+	 * @var FileContent
+	 */
+	private $contentOfWizardFile = null;
+
 	/**
 	 * WebUIFirstrunwizardContext constructor.
 	 *
 	 * @param FirstrunwizardPage $firstrunwizardPage
 	 */
+
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 11]
```diff
@@ -100,6 +112,56 @@ public function closeDialogBox() {
...
+	public function theHeadingOfThePopupShouldBe($expectedMessage) {
+		$headingMessage = $this->firstRunWizardPage->getHeadingMessage();
+
+		PHPUnit\Framework\Assert::assertSame(
+			$expectedMessage, $headingMessage, "Firstrunwizard showed unexpected message."
+		);
+	}
+
+	/**
+	 *@Given the administrator has changed the default popup message of firstrunwizard to :newMessage
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	 * @Given the administrator has changed the default popup message of firstrunwizard to :newMessage
```

File: tests/acceptance/features/lib/FirstrunwizardPage.php [hunk 12]
```diff
@@ -51,4 +53,26 @@ public function closePopup() {
 	public function getWizardPopup() {
 		return $this->find("xpath", $this->firstrunwizardPopupXpath);
 	}
+
+	/**
+	 * @return NodeElement|null
+	 */
+	public function clickShowFirstRunWizardPopupAgain() {
+		$openFirstRunwizardPopUpAgain = $this->find("xpath", $this->showFirstRunwizardScreenAgainButtonXpath);
+		$this->assertElementNotNull($openFirstRunwizardPopUpAgain, "No 'Show First Run Wizard Again' button");
```
Comment (paurakhsharma) [inline tests/acceptance/features/lib/FirstrunwizardPage.php]:
```suggestion
		$this->assertElementNotNull(
			$openFirstRunwizardPopUpAgain,
			__METHOD__ .
			" xpath $this->showFirstRunwizardScreenAgainButtonXpath could not find 'Show First Run Wizard Again'button"
		);
```

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+		return $this->find("xpath", $this->firstRunWizardXpath);
+	}
+
+	/**
+	 * Shows popup message from firstrunwizard in Settings page
+	 *
+	 * @return NodeElement|null
+	 */
+	public function showFirstRunWizardPopupInSettingsPage() {
+		$openFirstRunwizardPopUpAgain = $this->find("xpath", $this->requestToShowFirstRunwizardScreenInSettingsPageXpath);
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]:
```suggestion
		$firstrunwizardPopUp = $this->find("xpath", $this->requestToShowFirstRunwizardScreenInSettingsPageXpath);
```

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Page;
+
+use \Behat\Mink\Element\NodeElement;
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]: @bhawanaprasain Do we need `\` here at the front?

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Page object for General section of Settings page
+ *
+ */
+
+class PersonalGeneralSettingsPageFirstRunWizard extends PersonalGeneralSettingsPage {
+	protected $closeWizardXpath = "//*[@id='closeWizard']";
+	protected $firstRunWizardXpath = "//*[@id='firstrunwizard']";
+	protected $requestToShowFirstRunwizardScreenInSettingsPageXpath ="//*[@id=\"showWizard\"]";
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]: Too long variable name

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	/**
+	 * Shows popup message from firstrunwizard in Settings page
+	 *
+	 * @return NodeElement|null
+	 */
+	public function showFirstRunWizardPopupInSettingsPage() {
+		$openFirstRunwizardPopUpAgain = $this->find("xpath", $this->requestToShowFirstRunwizardScreenInSettingsPageXpath);
+		$this->assertElementNotNull($openFirstRunwizardPopUpAgain, "No 'Show First Run Wizard Again' button");
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]: Better error message would be better:
```
'Show First Run Wizard' button could not be found.
```

File: tests/acceptance/features/lib/FirstrunwizardPage.php [hunk 17]
```diff
@@ -33,6 +33,7 @@ class FirstrunwizardPage extends OwncloudPage {
 	protected $closeWizardXpath = "//*[@id='closeWizard']";
 	protected $headingMessageXpath = "//*[@id='firstrunwizard']/h1";
 	protected $firstRunWizardXpath = "//*[@id='firstrunwizard']";
+
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstrunwizardPage.php]: :+1: to this. I had forgotten.

**Comment by bhawanaprasain** [comment]
@dpakach @skshetry @paurakhsharma please review again

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 18]
```diff
@@ -44,9 +46,14 @@ class WebUIFirstrunwizardContext extends RawMinkContext implements Context {
 	private $webUIGeneralContext;
 
 	/**
-	 * @var FirstrunwizardPage
+	 * @var FirstRunWizardPage
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	 * @var FirstRunWizardPage $firstRunWizardPage
```

File: tests/acceptance/features/lib/FirstRunWizardPage.php [hunk 19]
```diff
@@ -49,6 +51,13 @@ public function closePopup() {
 		$this->waitTillElementIsNull($this->closeWizardXpath);
 	}
 
+	public function waitTillPageIsLoaded(
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstRunWizardPage.php]: Please add docstring

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+	 *
+	 * @return NodeElement|null
+	 */
+	public function showFirstRunWizardPopupInSettingsPage(Session $session) {
+		$firstRunWizardPopUp = $this->find("xpath", $this->showFirstRunWizardScreenXpath);
+		$this->assertElementNotNull($firstRunWizardPopUp, "'Show First Run Wizard' button could not be found.");
+		$firstRunWizardPopUp->click();
+		/**
+		 *
+		 * @var FirstRunWizardPage $firstRunWizardPage
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]: :+1: on that. But, can you remove the line just before this? Would look better, and will be diff-friendly.

File: tests/acceptance/features/lib/FirstRunWizardPage.php [hunk 21]
```diff
@@ -22,17 +22,19 @@
 
 namespace Page;
 
+use Behat\Mink\Session;
 use \Behat\Mink\Element\NodeElement;
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstRunWizardPage.php]: Not related to your PR, but can you also remove `\` here?

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 22]
```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * ownCloud
+ *
+ * @author Bhawana Prasain  <bhawana.prs@gmail.com>
+ * @copyright Copyright (c)  Bhawana Prasain 2019 bhawana.prs@gmail.com
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]:
```suggestion
 * @copyright Copyright (c) Bhawana Prasain 2019 bhawana.prs@gmail.com
```

File: tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+class PersonalGeneralSettingsPageFirstRunWizard extends PersonalGeneralSettingsPage {
+	protected $showFirstRunWizardScreenXpath ="//*[@id=\"showWizard\"]";
+
+	/**
+	 * Shows popup message from firstrunwizard in Settings page
+	 *
+	 * @param Session $session
+	 *
+	 * @return NodeElement|null
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPageFirstRunWizard.php]: This does not return anything, does it?

**Comment by phil-davis** [comment]
@bhawanaprasain please pull the latest master, rebase and resolve conflicts.
Then you will get php_codesniffer which will point out acceptance test code style issues:
```
make test-php-style
```

**Comment by bhawanaprasain** [comment]
@dpakach @skshetry @paurakhsharma please review again

**Pull request merged**